File: 2025-1/sesion-de-clase-18/Persona.cpp
/* 
 * File:   Persona.cpp
 * Author: erichuiza
 * 
 * Created on May 26, 2025, 4:29 PM
 */

#include "Persona.h"

Persona::Persona() : Registro() {
    this->nombre = nullptr;
}

Persona::Persona(int dni, const char* nombre) : Registro() {
    this->dni = dni;
    this->setNombre(nombre);
}

Persona::Persona(const Persona& orig) : Registro(orig) {
    *this = orig;
}

int Persona::getDni() const {
    return dni;
}

void Persona::setDni(int dni) {
    this->dni = dni;
}

char* Persona::getNombre() const {
    return this->nombre;
}

void Persona::setNombre(const char* nombre) {
    if (this->nombre) {
        delete[] this->nombre;
    }
    
    this->nombre = new char[strlen(nombre) + 1];
    strcpy(this->nombre, nombre);
}

bool Persona::cargarDatos(ifstream& archivo) {
    int dni;
    if (!archivo.eof() && archivo >> dni) {
        archivo.ignore();
        char *nombre = leerCadena(archivo);
        
        this->setDni(dni);
        this->setNombre(nombre);
        
        return true;
    }

    return false;
}

Persona& Persona::operator=(const Persona& persona) {
    this->setDni(persona.getDni());
    this->setNombre(persona.getNombre());
    
    return *this;
}

void Persona::imprimir(ostream& os) const {
    os << left << setw(15) << this->dni
        << setw(50) << this->nombre;
}

Persona::~Persona() {
    delete[] nombre;
}

File: 2025-2/sesion-de-clase-25/src/Persona.java
/**
 *
 * @author erichuiza
 */
public abstract class Persona {
    private int dni;
    private String nombre;
    private int edad;

    public Persona(int dni, String nombre, int edad) {
        this.dni = dni;
        this.nombre = nombre;
        this.edad = edad;
    }
    
    public int getDni() {
        return dni;
    }

    public void setDni(int dni) {
        this.dni = dni;
    }

    public String getNombre() {
        return nombre;
    }

    public void setNombre(String nombre) {
        this.nombre = nombre;
    }

    public int getEdad() {
        return edad;
    }

    public void setEdad(int edad) {
        this.edad = edad;
    }

    @Override
    public String toString() {
        return String.format("DNI: %d, Nombre: %s, edad: %d, ", dni, nombre, edad);
    }
}
